File: src/HeatShield.cpp
/*
  API for the HeatShield hardware.
  
  The file is a part of the application programming interface for
  the HeatShield didactic tool for control engineering and 
  mechatronics education.
  
  This code is part of the AutomationShield hardware and software
  ecosystem. Visit http://www.automationshield.com for more
  details. This code is licensed under a Creative Commons
  Attribution-NonCommercial 4.0 International License.

  Created by Gergely Takács and Richard Köplinger. 
  Last update: 10.10.2018.
*/

#include "HeatShield.h"
#include "AutomationShield.h"
// Initializes hardware pins
void HeatShieldClass::begin(){  
      pinMode(HEAT_YPIN, INPUT);  		// Thermistor pin
      pinMode(HEAT_UPIN, OUTPUT); 		// Cartridge
}

// Writes PWM signal to Cartridge, input is in %
void HeatShieldClass::actuatorWrite(float percent){  
      analogWrite(HEAT_UPIN,AutomationShield.percToPwm(percent)); // Write to actuator
}

// Calculates equivalent temperature by computing the
// Steinhart-hart equation in the simplified Beta parameter
// form
float HeatShieldClass::sensorRead() {
    return (1 / ((1 / REF_TEMP) + (log(getThermistorResistance() / NTC_RES) / NTC_BETA))) - ABSZERO;    
}

// Reads thermistor voltage
float HeatShieldClass::getThermistorVoltage() { 
      return (float)analogRead(HEAT_YPIN) * ARES;
}

// Reads thermistor resistance
// Voltage divider circuit (Kirchhoffs law)
float HeatShieldClass::getThermistorResistance() {
	float Vterm = getThermistorVoltage();
	 return ((Vterm*VD_RES)/(VD_REF-Vterm));
}

HeatShieldClass HeatShield;

File: matlab/HeatShield.m
%   HeatShield MATLAB API.
%
%   The script initializes the board, then reads the
%   output, writes to the input and shows how alternative
%   readings from the termocouple work.
%
%   This code is part of the AutomationShield hardware and software
%   ecosystem. Visit http://www.automationshield.com for more
%   details. This code is licensed under a Creative Commons
%   Attribution-NonCommercial 4.0 International License.
%
%   Created by Gergely Takács.
%   Last update: 10.7.2019.

classdef HeatShield < handle

    properties
        arduino;
    end

    properties(Constant)
        HEAT_YPIN = 'A0';         % NTC Thermistor (Sensor)
        HEAT_UPIN = 'D3';         % Cartridge (Actuator)
        REF_TEMP = 25.0 + 273.15; % Thermistor reference temperature
        ABSZERO = 273.15;         % Absolute zero
        NTC_RES = 100000.0;       % Resistance of the thermistor
        VD_RES = 100000.0;        % Resistance of voltage divider arm
        VD_REF = 5.0;             % Input for the voltage divider
        NTC_BETA = 3950.0;        % value of Beta factor from datasheet MF58
    end

    methods
        function begin(HeatShieldObject)          % Initialization function
            HeatShieldObject.arduino = arduino();
            disp('HeatShield initialized.')
        end

        function actuatorWrite(HeatShieldObject, percent)
            writePWMDutyCycle(HeatShieldObject.arduino, HeatShieldObject.HEAT_UPIN, (percent / 100));
        end

        function U = getThermistorVoltage(HeatShieldObject)
            U = readVoltage(HeatShieldObject.arduino, HeatShieldObject.HEAT_YPIN);
        end

        function R = getThermistorResistance(HeatShieldObject)
            Vterm = HeatShieldObject.getThermistorVoltage();
            R = ((Vterm * HeatShieldObject.VD_RES) / (HeatShieldObject.VD_REF - Vterm));
        end

        function y = sensorRead(HeatShieldObject)
            y = (1 / ((1 / HeatShieldObject.REF_TEMP) + (log(HeatShieldObject.getThermistorResistance()/HeatShieldObject.NTC_RES) / HeatShieldObject.NTC_BETA))) - HeatShieldObject.ABSZERO;
        end
    end
end
